test/base/test_history.py

```diff
@@ -394,6 +430,42 @@ def test_trace_all_aggregated(self):
         self.check_history()
 
 
+class CRFunModeHistoryTest(HistoryTest):
```

Comment:
To me it is not clear, where the difference to `CRResModeHistoryTest` is, isn't this the exactly same code? And also it is not clear to me, where it is decided whether to use `ModeFun` or `ModeRes`?

Reply:
the difference is the used optimizer -- once it will call residuals, once function values.